examples/structured_data/deep_neural_decision_forests.py

```diff
@@ -0,0 +1,447 @@
+
+
+def encode_inputs(inputs, use_embedding=False):
+    encoded_features = []
+    for feature_name in inputs:
+        if feature_name in CATEGORICAL_FEATURE_NAMES:
+            vocabulary = CATEGORICAL_FEATURES_WITH_VOCABULARY[feature_name]
+            # Create a lookup to convert a string values to an integer indices.
+            # Since we are not using a mask token nor expecting any out of vocabulary
+            # (oov) token, we set mask_token to None and  num_oov_indices to 0.
+            index = StringLookup(
+                vocabulary=vocabulary, mask_token=None, num_oov_indices=0
+            )
+            # Convert the string input values into integer indices.
+            value_index = index(inputs[feature_name])
+            if use_embedding:
+                embedding_dims = int(math.sqrt(len(vocabulary)))
+                # Create an embedding layer with the specified dimensions.
+                embedding_ecoder = layers.Embedding(
+                    input_dim=len(vocabulary), output_dim=embedding_dims
+                )
+                # Convert the index values to embedding representations.
+                encoded_feature = embedding_ecoder(value_index)
+            else:
+                # Create a one-hot encoder.
+                onehot_encoder = CategoryEncoding(output_mode="binary")
+                onehot_encoder.adapt(index(vocabulary))
+                # Convert the index values to a one-hot representation.
+                encoded_feature = onehot_encoder(value_index)
+        else:
+            # Use the numerical features as-is.
+            encoded_feature = inputs[feature_name]
+            if inputs[feature_name].shape[-1] is None:
+                encoded_feature = tf.expand_dims(encoded_feature, -1)
+
+        encoded_features.append(encoded_feature)
+
+    encoded_features = layers.concatenate(encoded_features)
+    return encoded_features
+
+
+"""
+## Deep Neural Decision Tree
+
+A neural decision tree model has two sets of weights to learn. The first set is `pi`,
+which represents the probability distribution of the classes in the tree leaves.
+The second set is the weights of the routing layer `decision_fn`, which represents the probability
+of going to each leave. The forward pass of the model works as follows:
+
+1. The model expects input `features` as a single vector encoding all the features of an instance
+in the batch. This vector could be generated from a Convolution Neural Network (CNN) applied to images,
+or dense transformations applied to structured data features. 
+2. The model first applies a `used_features_mask` to randomly select a subset of input features to use.
+3. Second, the model computes the probabilities `mu` for the input instances to reach the tree leaves,
+by iteratively performing a *stochastic* routing throughout the tree levels.
+4. Finally, the probabilities of reaching the leaves are combined by the class probabilities at the
+leaves to produce the final `outputs`.
```

Comment:
Nit (step 3 is not "second"):

```suggestion
1. The model expects input `features` as a single vector encoding all the features of an instance
in the batch. This vector can be generated from a Convolution Neural Network (CNN) applied to images
or dense transformations applied to structured data features. 
2. The model first applies a `used_features_mask` to randomly select a subset of input features to use.
3. Then, the model computes the probabilities (`mu`) for the input instances to reach the tree leaves
by iteratively performing a *stochastic* routing throughout the tree levels.
4. Finally, the probabilities of reaching the leaves are combined by the class probabilities at the
leaves to produce the final `outputs`.
```